peachjam/signals.py

```diff
@@ -75,3 +75,14 @@ def doc_deleted_update_extracted_citations(sender, instance, **kwargs):
     """Update language list on related work after a subclass of CoreDocument is deleted."""
     if isinstance(instance, CoreDocument):
         update_extracted_citations_for_a_work(instance.work_id)
+
+
+@receiver(signals.post_save)
```

Comment:
What if we approach this a bit differently to avoid the circular logic issue.

Instead of queuing this up every time a document is saved, what if we explicitly queue it up in the admin form when a document is created/edited in the admin area, or if it's created/updated via an ingestor?

My motivation is that lots of other background logic (including citation extraction) saves a document, but that doesn't always mean we need to re-extract citations of future documents. We only actually want to do re-extract citations if the document is brand new, or has substantially changed. Let's assume that editing in the admin area means "substantially changed", and not background task updates.

Or, slightly differently, we use the signal handler to only handle the case where a document is created (ie. check for the creation flag in the handler), and then in the admin form we handle the "document changed" case.

That would allow us to remove the (rather odd) explicit flag. The flag is awkward because we're decorating the document to communicate to the signal handler.

Reply:
@longhotsummer what if in the signal we check whether the document is new, or check only if the date has changed since that is the value we are most interested in?

Reply:
If we're going to restrict it to certain fields, that must be done in the Admin area because then we have access to the form which can tell us what fields changed. We'd need to ensure that the fields we monitor are those that are checked during citation matching. I think these are: alternative names, citation, title (and date).